mission_control/navigator_missions/navigator_missions/entrance_gate.py

```diff
@@ -0,0 +1,463 @@
+                continue
+
+            # Determine which gate the intersection is in
+            bucket_index = np.argmin(np.array(gateDist))
+            bucket[bucket_index] = bucket[bucket_index] + 1
+
+        # Determine which gate has the most hits
+        pinger_gate = np.argmax(np.array(bucket))
+
+        # Calculate the confidence
+        confidence = bucket[pinger_gate] / (bucket[0] + bucket[1] + bucket[2])
+        return (pinger_gate, confidence)
+
+    @util.cancellableInlineCallbacks
+    def ping_recv(self, p_message):
+        try:
+            # Transform the ping into enu
+            hydrophones_to_enu = yield self.tf_listener.get_transform('enu', p_message.header.frame_id)
+            hydrophones_origin = hydrophones_to_enu._p[0:2]
+            heading = rosmsg_to_numpy(p_message.vector)
+            heading_enu = hydrophones_to_enu.transform_vector(heading)
+            heading_enu = heading_enu[0:2] / np.linalg.norm(heading_enu[0:2])
+
+            # Track the ping
+            self.intersect_vectors.append((hydrophones_origin[0:2], hydrophones_origin[0:2] + heading_enu[0:2]))
+        except tf2_ros.TransformException, e:
+            rospy.logwarn('TF Exception: {}'.format(e))
+
+    '''
+
+        Multilaterating
+
+    '''
+
+    @util.cancellableInlineCallbacks
+    def get_multilateration_closest(self):
+        # Get the most recent multilateration position
+        rospos = yield self.hydrophones.get_last_position()
+        pinger_pos = rosmsg_to_numpy(rospos.point)[0:2]
+
+        # Calculate the distances to each gate center
+        distances = []
+        for gate_center in self.gate_centers:
+            distances.append((gate_center[0] - pinger_pos[0]) ** 2 + (gate_center[1] - pinger_pos[1]) ** 2)
+
+        # Determine which gate is closest
+        pinger_gate = np.argmin(np.array(distances))
+
+        # Calculate the confidence
+        confidence = (5 - distances[pinger_gate]) / 5.0
+        defer.returnValue((pinger_gate, confidence))
+
+    '''
+
+        Math Utilities
+
+    '''
+
+    @staticmethod
+    def line(p1, p2):
+        '''
+        Return equation of a line given two 2D points
+        https://stackoverflow.com/questions/20677795/how-do-i-compute-the-intersection-point-of-two-lines-in-python
+        '''
+        A = (p1[1] - p2[1])
+        B = (p2[0] - p1[0])
+        C = (p1[0] * p2[1] - p2[0] * p1[1])
+        return A, B, -C
+
+    @staticmethod
+    def perpendicular(L):
+        '''
+        Calculate the line perpendicular to another line
+        '''
+        v1 = [L[1], -L[0], 0]
+        r = v1[0] ** 2 + v1[1] ** 2
+        v1 = [v1[0] / r, v1[1] / r]
+        v2 = [0, 0, 1]
+        pvec = np.cross(v1, v2)
+        r = math.sqrt(pvec[0] ** 2 + pvec[1] ** 2)
+        return [pvec[0] / r, pvec[1] / r, 0]
+
+    '''
+
+        Perception Utilities
+
+    '''
+
+    @util.cancellableInlineCallbacks
+    def find_nearest_objects(self, name, number):
```

Comment:
To maximize code reuse, perhaps you could instead use the function added in #443